File: sdk/symbol_dart_sdk/bin/ByteArray.dart
import 'dart:typed_data';
import 'package:convert/convert.dart';

import 'utils/converter.dart';

class ByteArray {
  Uint8List bytes;

  ByteArray(int fixedSize, dynamic arrayInput)
      : bytes = Uint8List(fixedSize) {
    var rawBytes = arrayInput;
    if (rawBytes is String) {
      try {
        if (isHexString(rawBytes)) {
          rawBytes = hex.decode(rawBytes); // Remove '0x' prefix
        } else {
          rawBytes = stringToAddress(arrayInput);
        }
      } catch (e) {
        throw ArgumentError('bytes was not a valid hex or address string');
      }
    }

    if (fixedSize != rawBytes.length) {
      throw RangeError('bytes was size ${rawBytes.length} but must be $fixedSize');
    }

    bytes = Uint8List.fromList(rawBytes);
  }

  @override
  String toString() {
    try {
      return addressToString(bytes);
    } catch(_) {
      return hex.encode(bytes).toUpperCase();
    }
  }
}

File: sdk/csharp/SymbolSdk/ByteArray.cs
namespace SymbolSdk;
/**
 * Represents a fixed size byte array.
 */
public abstract class ByteArray
{
    public uint Size { get; }
    public byte[] bytes { get; }

    /**
	 * Creates a byte array.
	 * @param {byte} fixedSize Size of the array.
	 * @param {byte[]?} arrayInput Byte array.
	 */
    protected ByteArray(byte fixedSize, object? arrayInput)
    {
        bytes = new byte[fixedSize];
        Size = fixedSize;
        switch (arrayInput)
        {
            case byte[] bytes1:
            {
                if (fixedSize != bytes1.Length) throw new Exception($"bytes was size {bytes1.Length} but must be {fixedSize}");
                bytes = bytes1;
                break;
            }
            case string {Length: 39 or 40} str:
                bytes = Converter.StringToAddress(str);
                break;
            case string str when Converter.IsHexString(str):
            {
                var rawBytes = Converter.HexToBytes(str);
                if (fixedSize != rawBytes.Length) throw new Exception($"bytes was size {rawBytes.Length} but must be {fixedSize}");
                bytes = rawBytes;
                break;
            }
            case string str:
                throw new Exception("Invalid input type.");
        }
    }
    
    /**
	 * Returns string representation of this object.
	 * @returns {string} String representation of this object
	 */
    public override string ToString()
    {
        if (GetType() == typeof(Symbol.UnresolvedAddress) || GetType() == typeof(Symbol.Address) || GetType() == typeof(Symbol.UnresolvedAddress))
        {
            return Converter.AddressToString(bytes);
        }

        if (GetType() == typeof(Nem.Address))
        {
            return Converter.BytesToUtf8(bytes);
        }
        return Converter.BytesToHex(bytes);
    }
}
